smartsim/entity/application.py

```diff
@@ -0,0 +1,288 @@
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from __future__ import annotations
+
+import copy
+import typing as t
+from os import path as osp
+
+from .._core.utils.helpers import expand_exe_path
+from ..log import get_logger
+from .entity import SmartSimEntity
+from .files import EntityFiles
+
+if t.TYPE_CHECKING:
+    from smartsim.types import TODO
+
+    RunSettings = TODO
+    BatchSettings = TODO
+
+
+logger = get_logger(__name__)
+
+
+# TODO: Remove this supression when we strip fileds/functionality
+#       (run-settings/batch_settings/params_as_args/etc)!
+# pylint: disable-next=too-many-public-methods
+
+
+class Application(SmartSimEntity):
+    def __init__(
+        self,
+        name: str,
+        exe: str,
+        exe_args: t.Optional[t.Union[str, t.Sequence[str]]] = None,
+        files: t.Optional[EntityFiles] = None,
+        file_parameters: t.Mapping[str, str] | None = None,
+    ) -> None:
+        """Initialize an ``Application``
+
+        :param name: name of the application
+        :param exe: executable to run
+        :param exe_args: executable arguments
+        :param files: files to be copied, symlinked, and/or configured prior to
+                      execution
+        :param file_parameters: parameters and values to be used when configuring
+                                files
+        """
+        super().__init__(name)
+        self._exe = expand_exe_path(exe)
+        self._exe_args = self._build_exe_args(exe_args) or []
+        self._files = copy.deepcopy(files) if files else None
+        self._file_parameters = (
+            copy.deepcopy(file_parameters) if file_parameters else {}
+        )
+        self._incoming_entities: t.List[SmartSimEntity] = []
+        self._key_prefixing_enabled = False
+
+    @property
+    def exe(self) -> str:
+        """Return executable to run.
+
+        :returns: application executable to run
+        """
+        return self._exe
+
+    @exe.setter
+    def exe(self, value: str) -> None:
+        """Set executable to run.
+
+        :param value: executable to run
+        """
+        self._exe = copy.deepcopy(value)
+
+    @property
+    def exe_args(self) -> t.Sequence[str]:
+        """Return an immutable list of attached executable arguments.
```

Comment:
At runtime, not technically, but the type hint here states that a sequence is return which implicitly promotes immutability. If we do drop the "immutable" wording from the docstring, I'm going to also recommend changing the return type to `list[str]` or at the very least `MutableSequence[str]`

Reply:
Yeah, I guess I was thinking that ``_build_exe_args`` returns a ``List`` which is mutable so that part was out of sync.  

Do we want to be specific and risk API breaks later by changing to ``list[str]`` or generic to limit breaks later (i.e. ``MutableSequence[str]``)

Reply:
changed to mutablesequence[str]